stratum/hal/lib/tdi/es2k/es2k_sde_meter.cc

```diff
@@ -176,8 +176,7 @@ ::util::Status GetMeterField(
     RETURN_IF_TDI_ERROR(table_data->getValue(field_id, &redPackets));
     cfg.redPackets = redPackets;
   } else {
-    MAKE_ERROR(ERR_INVALID_PARAM)
-        << "Unknown meter field " << field_name << " in meter table ";
```

Comment:
POLICER_METER_SPEC_VALID
POLICER_METER_SPEC_ERROR
POLICER_METER_SPEC_PF_NUM
POLICER_METER_SPEC_HOST
POLICER_METER_SPEC_FUNC_TYPE
POLICER_METER_SPEC_FUNC_NUM
POLICER_METER_SPEC_CTOK
POLICER_METER_SPEC_ETOK
POLICER_METER_SPEC_TIME_STORED

These are the meter fields that are defined in tdi.json(earlier bf-rt.json) which are not defined in tdi_constants.h.
Even though we define all these constants in tdi_constants.h, it will still not solve the problem.
Problem would be solved only if we add additional checks for doing a get operation on all these unused fields in GetMeterField() method.
But do we really need to fetch such unused meter fields in our code? The answer is no. If we don't need it, why fetch it.
Tomorrow if some new additional fields are added in tdi.json, will we come back again to update the GetMeterField() method, just to avoid unnecessary error logging, even though we don't really use such meter fields in our codebase? IMO this is not the right way.

To answer Derek's question, if it is a bug. It's not a bug. Infact it is a cosmetic error.
As an analogy, it's like "I asked you to fetch apples. You say here are the apples, but I didn't find the oranges and you log not finding oranges as an error".
But was that the requirement in the first place? Logging such errors create an impression that oranges were needed, which is not true.

It makes perfect sense to add RETURN_IF_TDI_ERROR in all the conditions where we actually expect valid values since we use those meter fields and if we are not able to correctly fetch these values, it's an error and we should treat it like that. But for meter fields which we don't need, which we don't use, why should we even bother fetching it, why should we even log error if we are not able to fetch it?
  
P.S: Usage of MAKE_ERROR in any form creates an impression that it is an error. I didn't want to completely ignore the else check, and wanted to log this as a warning, so that the developer gets that ceratin meter fields are unknown/unsupported for us.

Reply:
One more question before I reply further: if we get one call to access an undefined field, are we likely to get many?

Reply:
Also, when I tested `return MAKE_ERROR().severity(WARNING)`, I didn't see any difference in the message, which didn't include the word `ERROR`.

I inserted the statement in the `main` module, to reduce the amount of setup needed to trigger the message, but I can't think of any reason that would make a difference.

Reply:
> If we get one call to access an undefined field, are we likely to get many?
> 
> Ah. I may already have the answer. From the first part of your reply, it would appear that we are already getting warnings on the nine fields you mention. Is that correct?

Yes.
infrap4d logs:
E20231130 18:56:58.132731  2124 status_macros.cc:128] MakeErrorStream destructed without getting Status: /sdk/build/hw/p4-cp-nws/stratum/stratum/stratum/hal/lib/tdi/es2k/es2k_sde_meter.cc:180 Unknown meter field $POLICER_METER_SPEC_VALID in meter table
E20231130 18:56:58.132781  2124 status_macros.cc:128] MakeErrorStream destructed without getting Status: /sdk/build/hw/p4-cp-nws/stratum/stratum/stratum/hal/lib/tdi/es2k/es2k_sde_meter.cc:180 Unknown meter field $POLICER_METER_SPEC_ERROR in meter table
E20231130 18:56:58.132798  2124 status_macros.cc:128] MakeErrorStream destructed without getting Status: /sdk/build/hw/p4-cp-nws/stratum/stratum/stratum/hal/lib/tdi/es2k/es2k_sde_meter.cc:180 Unknown meter field $POLICER_METER_SPEC_PF_NUM in meter table
E20231130 18:56:58.132814  2124 status_macros.cc:128] MakeErrorStream destructed without getting Status: /sdk/build/hw/p4-cp-nws/stratum/stratum/stratum/hal/lib/tdi/es2k/es2k_sde_meter.cc:180 Unknown meter field $POLICER_METER_SPEC_HOST in meter table
E20231130 18:56:58.132829  2124 status_macros.cc:128] MakeErrorStream destructed without getting Status: /sdk/build/hw/p4-cp-nws/stratum/stratum/stratum/hal/lib/tdi/es2k/es2k_sde_meter.cc:180 Unknown meter field $POLICER_METER_SPEC_FUNC_TYPE in meter table
E20231130 18:56:58.132843  2124 status_macros.cc:128] MakeErrorStream destructed without getting Status: /sdk/build/hw/p4-cp-nws/stratum/stratum/stratum/hal/lib/tdi/es2k/es2k_sde_meter.cc:180 Unknown meter field $POLICER_METER_SPEC_FUNC_NUM in meter table
E20231130 18:56:58.132856  2124 status_macros.cc:128] MakeErrorStream destructed without getting Status: /sdk/build/hw/p4-cp-nws/stratum/stratum/stratum/hal/lib/tdi/es2k/es2k_sde_meter.cc:180 Unknown meter field $POLICER_METER_SPEC_CTOK in meter table
E20231130 18:56:58.132869  2124 status_macros.cc:128] MakeErrorStream destructed without getting Status: /sdk/build/hw/p4-cp-nws/stratum/stratum/stratum/hal/lib/tdi/es2k/es2k_sde_meter.cc:180 Unknown meter field $POLICER_METER_SPEC_ETOK in meter table
E20231130 18:56:58.132881  2124 status_macros.cc:128] MakeErrorStream destructed without getting Status: /sdk/build/hw/p4-cp-nws/stratum/stratum/stratum/hal/lib/tdi/es2k/es2k_sde_meter.cc:180 Unknown meter field $POLICER_METER_SPEC_TIME_STORED in meter table

Reply:
Also, I think MAKE_ERROR() usage in any form is going to emit the error msg on infrap4d console. I tried with .without_logging() attribute, but was still getting those msgs on the console.

Reply:
In other words, we _expect_ to be called with fields we're not interested in.

The correct solution then is:

1. Add a comment stating our intention.
2. Don't log at all.

I would do this by adding a comment at the top of the `if`/`else` ladder:

```c
  // We only return a value if field_name specifies one of the meter
  // fields we're interested in. If it isn't, just fall through and
  // return OK.
  if (field_name == kEs2kMeterProfileIdKPps) {
    uint64 prof_id;
    RETURN_IF_TDI_ERROR(table_data->getValue(field_id, &prof_id));
    cfg.meter_prof_id = prof_id;
    cfg.isPktModMeter = true;
  } else if (field_name == kEs2kMeterCirPps) {
```

and eliminating the final `else`:

```c
  } else if (field_name == kEs2kMeterRedCounterPackets) {
    uint64 redPackets;
    RETURN_IF_TDI_ERROR(table_data->getValue(field_id, &redPackets));
    cfg.redPackets = redPackets;
  }

  return ::util::OkStatus();
```

Reply:
### Takeaways

#### 1. Write good changelist descriptions

To quote from the Google code review practices, [Writing good CL descriptions](https://google.github.io/eng-practices/review/developer/cl-descriptions.html):

"The first line should be a short, focused summary, while the rest of the description should fill in the details and include any supplemental information a reader needs to understand the changelist holistically. It might include a brief description of the problem that’s being solved, and why this is the best approach. If there are any shortcomings to the approach, they should be mentioned. If relevant, include background information such as bug numbers, benchmark results, and links to design documents...


"Even small CLs deserve a little attention to detail. Put the CL in context."

In this case, the CL comment did not provide enough context (a clear statement of the problem) for the reviewers to evaluate the PR.

#### Log with intent

One of the biggest problems with logs is signal-to-noise ratio. The more stuff you write -- the more noise there is -- the harder it is to pick out the stuff that matters.

Don't log unless you have something useful to say. If you do log, provide the user with the information they need to understand and address the issue.

This is an expected condition, not an exceptional one. And it's one that's going to occur frequently, and in bursts. It qualifies as noise.

#### 3. Leave an explanation for the maintainer

"A code review is often the first test of whether a given change is understandable to a broader audience."
-- _Software Engineering at Google_, p. 172

If the code does something that runs against common sense, make it clear that you are doing it intentionally, and explain why you're doing it or why what you're doing isn't a problem.

The most useful thing you can do in a comment is answer the question _why?_